test/core/event_engine/posix/poller_fork_test.cc

```diff
@@ -0,0 +1,415 @@
+  ASSERT_THAT(client.status(), IsOk());
+  auto endpoint = AwaitEndpoint();
+  // Start read and write, cause the fork. Both operations should succeed
+  // post-fork.
+  StatusListener read_status;
+  StatusListener write_status;
+  SliceBuffer read_buffer;
+  SliceBuffer write_buffer;
+  // 4M seems to be enough to fill the buffers on my Linux instance. May need to
+  // be adjusted in the future!
+  MutableSlice slice = MutableSlice::CreateUninitialized(4 * 1024 * 1024);
+  std::fill(slice.begin(), slice.end(), 42);
+  write_buffer.Append(Slice(std::move(slice)));
+  ASSERT_FALSE(endpoint->Read(read_status.Setter(), &read_buffer, nullptr));
+  ASSERT_FALSE(endpoint->Write(write_status.Setter(), &write_buffer, nullptr))
+      << "Need to send more data";
+  LOG(INFO) << "Before fork in parent";
+  // Let the data reach the buffers
+  absl::SleepFor(absl::Milliseconds(50 * grpc_test_slowdown_factor()));
+  ee()->BeforeForkForTests();
+  ee()->AfterForkForTests(false);
+  LOG(INFO) << "After fork in parent";
+  ASSERT_THAT(client.Read(write_buffer.Length()), IsOk());
+  ASSERT_THAT(client.Write("Hi!"), IsOk());
+  EXPECT_THAT(read_status.AwaitStatus(), IsOk());
+  EXPECT_THAT(write_status.AwaitStatus(), IsOk());
+  // Starting read and write post-fork will fail asynchronously and return the
+  // status.
+  ASSERT_FALSE(endpoint->Read(read_status.Setter(), &read_buffer, nullptr));
+  ASSERT_THAT(client.Write("Hi again"), IsOk());
+  EXPECT_THAT(read_status.AwaitStatus(), IsOk());
+  bool write_result =
+      endpoint->Write(write_status.Setter(), &write_buffer, nullptr);
+  ASSERT_THAT(client.Read(write_buffer.Length()), IsOk());
+  if (!write_result) {
+    EXPECT_THAT(write_status.AwaitStatus(),
+                StatusIs(absl::StatusCode::kInternal));
+  }
+}
+
+TEST_F(PollerForkTest, ListenerInChild) {
+  // Connect before "fork"
+  RawPosixClient client(address_);
+  ASSERT_THAT(client.status(), IsOk());
+  auto endpoint = AwaitEndpoint();
+  // Start read and write
+  StatusListener read_status;
+  StatusListener write_status;
+  SliceBuffer read_buffer;
+  SliceBuffer write_buffer;
+  // 4M seems to be enough to fill the buffers on my Linux instance. May need to
+  // be adjusted in the future!
+  MutableSlice slice = MutableSlice::CreateUninitialized(4 * 1024 * 1024);
+  std::fill(slice.begin(), slice.end(), 42);
+  write_buffer.Append(Slice(std::move(slice)));
+  ASSERT_FALSE(endpoint->Read(read_status.Setter(), &read_buffer, nullptr));
+  ASSERT_FALSE(endpoint->Write(write_status.Setter(), &write_buffer, nullptr))
+      << "Need to send more data";
+  LOG(INFO) << "Before fork in child";
+  // Let the data reach the buffers
+  absl::SleepFor(absl::Milliseconds(50 * grpc_test_slowdown_factor()));
+  ee()->BeforeForkForTests();
+  ee()->AfterForkForTests(true);
+  LOG(INFO) << "After fork in child";
+  EXPECT_THAT(read_status.AwaitStatus(),
+              StatusIs(absl::StatusCode::kResourceExhausted));
```

Comment:
Why should this be ResourceExhausted? It seems like Unavailable would be a more appropriate status here. This may be an implementation detail that we don't want to hard-code in tests.